### ISSUE TITLE
cURL error 7 when trying to refresh the access token

### ISSUE THREAD

**Issue by SparkyTD** [googleapis/google-api-php-client#1985]
Hello! I am using this API client on my website, and I've ran into a problem. On my development server (Manjaro), everything works as expected, and I can complete the OAuth process. The problem is that on the deployment server (CentOS 8), the last step of the authentication, when Google returns the `code`  to my website so I can convert it into an access token, the function that does this conversion throws the following exception:

```

Fatal error: Uncaught GuzzleHttp\Exception\ConnectException: cURL error 7: (see https://curl.haxx.se/libcurl/c/libcurl-errors.html) for https://oauth2.googleapis.com/token 
in /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php:207 Stack trace: 
#0 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(167): GuzzleHttp\Handler\CurlFactory::createRejection() 
#1 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php(114): GuzzleHttp\Handler\CurlFactory::finishError() 
#2 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Handler/CurlHandler.php(56): GuzzleHttp\Handler\CurlFactory::finish() 
#3 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Handler/Proxy.php(28): GuzzleHttp\Handler\CurlHandler->__invoke() 
#4 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Handler/Proxy.php(51): GuzzleHttp\Handler\Proxy::GuzzleHttp\Handler\{closure}() 
#5 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/PrepareBodyMiddleware.php(66): GuzzleHttp\Handler\Proxy::GuzzleHttp\Handler\{closure}() 
#6 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Middleware.php(29): GuzzleHttp\PrepareBodyMiddleware->__invoke() 
#7 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/RedirectMiddleware.php(70): GuzzleHttp\Middleware::GuzzleHttp\{closure}() 
#8 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Middleware.php(57): GuzzleHttp\RedirectMiddleware->__invoke() 
#9 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/HandlerStack.php(71): GuzzleHttp\Middleware::GuzzleHttp\{closure}() 
#10 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Client.php(351): GuzzleHttp\HandlerStack->__invoke() 
#11 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Client.php(112): GuzzleHttp\Client->transfer() 
#12 /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Client.php(129): GuzzleHttp\Client->sendAsync() 
#13 /var/www/html/katalogus/vendor/google/auth/src/HttpHandler/Guzzle6HttpHandler.php(47): GuzzleHttp\Client->send() 
#14 /var/www/html/katalogus/vendor/google/auth/src/OAuth2.php(526): Google\Auth\HttpHandler\Guzzle6HttpHandler->__invoke() 
#15 /var/www/html/katalogus/vendor/google/apiclient/src/Client.php(247): Google\Auth\OAuth2->fetchAuthToken() 
#16 /var/www/html/katalogus/login.php(47): Google\Client->fetchAccessTokenWithAuthCode() 
#17 {master} thrown in /var/www/html/katalogus/vendor/guzzlehttp/guzzle/src/Handler/CurlFactory.php on line 207
```

I have tried reverse engineering the actual curl request that is being sent from the library so I can use it as a standalone command, and this is what I ended up with:

```
curl --request POST \
  --url https://oauth2.googleapis.com/token \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --data grant_type=authorization_code \
  --data code=[redacted] \
  --data redirect_uri=[redacted] \
  --data client_id=[redacted] \
  --data client_secret=[redacted]
```

When I execute this on the same deployment server, it works as expected and returns the authorization codes. This makes me believe that the problem is not with the internet connection or firewall, but rather in the Google API Client itself. The problem also seems to be specific to the deployment server since I haven't been able to reproduce it anywhere else.

#### Environment details

  - OS: CentOS 8
  - PHP version: 7.4.10 [dev/working server] / 7.4.12 [deployment/broken server]
  - cUrl version 7.72.0 [dev/working server] / 7.61.1 [deployment/broken server]
  - Package name and version: google/apiclient v2.8.1

#### Steps to reproduce

  1. Create an instance of `Google\Client`
  2. Set up all the parameters (client id, secret, scopes, etc...)
  3. Call `fetchAccessTokenWithAuthCode($code)` by completing the OAuth flow

#### Code example

```php
$g_client = new Google\Client();
$g_client->setClientId($google_clientID);
$g_client->setClientSecret($google_clientSecret);
$g_client->setRedirectUri($google_redirectUri);
$g_client->setPrompt("select_account");
$g_client->addScope("profile");
$g_client->addScope("https://www.googleapis.com/auth/admin.directory.group.readonly");
$g_client->addScope("https://www.googleapis.com/auth/admin.directory.user.readonly");
$g_client->addScope("https://www.googleapis.com/auth/userinfo.email");
$g_client->addScope("https://www.googleapis.com/auth/userinfo.profile");

// ...

// Handle the code response from Google
$accessToken = $g_client->fetchAccessTokenWithAuthCode($_GET['code']);
```

**Comment by SparkyTD** [comment]
Update: I have narrowed it down even further: The culprit is probably not the Google API client.
I wrote the following code snippet:
```
$conf = unserialize($ser_conf);
$handle = curl_init();
curl_setopt_array($handle, $conf);
$server_output = curl_exec($handle);
die(curl_errno($handle) . ">>>>> " . $server_output);
```
The variable $ser_conf is a string that was obtained by calling `serialize()` on the $conf object in `guzzle/src/Handler/CurlFactory.php:44`. If I run the snippet in a standalone php file with the serialized string, it works, but if I run the same snippet in the create() function in the guzzle file I mentioned, it fails with errno 7.
